cmd/snap/cmd_run.go

```diff
@@ -493,30 +483,115 @@ func (x *cmdRun) straceOpts() (opts []string, raw bool, err error) {
 	return opts, raw, nil
 }
 
+// isSnapRefreshConflictDetected detects if snap refreshed was started while not
+// holding the inhibition hint file lock.
+//
+// For context, on snap first install, the inhibition hint lock file is not created
+// so we cannot hold it. It is created after the first refresh. This allows for a
+// window where we don't hold the lock where a snap refresh could start.
+//
+// The following scenarios are detected:
+// 	- There is an ongoing refresh and the current symlink was removed temporarily in the process.
+// 	- A full refresh took place hence initial revision is not the same as the current revision.
+// 	- Snap was removed.
+func isSnapRefreshConflictDetected(oldInfo *snap.Info, app *snap.AppInfo, hintFlock *osutil.FileLock) bool {
+	// XXX: An alternative for this complex check is to check if hint file exists.
+	// It didn't exist before and exists now hence a race condition is detected.
+	if !features.RefreshAppAwareness.IsEnabled() || app.IsService() || hintFlock != nil {
+		// Skip check
+		return false
+	}
+
+	if !osutil.FileExists(filepath.Join(dirs.SnapMountDir, oldInfo.InstanceName(), "current")) {
+		// Current symlink does not exist, snap is either removed or being refreshed.
+		logger.Debugf("possible ongoing refresh conflict detected; current symlink does not exist")
+		return true
+	}
+
+	currentInfo, err := snap.ReadCurrentInfo(oldInfo.InstanceName())
```

Comment:
I've pasted ReadCurrentInfo here. I think we still have another funny issue:

```go
func ReadCurrentInfo(snapName string) (*Info, error) {
	curFn := filepath.Join(dirs.SnapMountDir, snapName, "current")
	realFn, err := os.Readlink(curFn)
	if err != nil {
		return nil, fmt.Errorf("cannot find current revision for snap %s: %s", snapName, err)
	}
	rev := filepath.Base(realFn)

	// XXX say we have revision 99 here

	revision, err := ParseRevision(rev)
	if err != nil {
		return nil, fmt.Errorf("cannot read revision %s: %s", rev, err)
	}

	// XXX snap removes current and it's pointing to something else, but we still read from
	// revision 99

	// XXX so snap info returned here will be true
	// in currentInfo.SnapRevision() != oldInfo.SnapRevision() check

	return ReadInfo(snapName, &SideInfo{Revision: revision})
}
```

this probably isn't much of a problem, and we'll likely run the old revision successfully.

Reply:
Nice catch, This definitely a race condition!
I had a suggestion earlier about another way to detect possible race condition without checking the snap and only checking that inhibition hint file exists https://github.com/snapcore/snapd/pull/13747#discussion_r1537469185 it should simplify the check and make it race free

Reply:
Thank for pointing it out, This commit should address this race condition https://github.com/snapcore/snapd/pull/13747/commits/51a6bd445145a4a3c579478aa4bb1de01bba4e60